src/signalfx_metadata.py

```diff
@@ -0,0 +1,255 @@
+    max_possible = 8  # initial value
+    while True:
+        bytes = max_possible * struct_size
+        names = array.array('B', '\0' * bytes)
+        outbytes = struct.unpack('iL', fcntl.ioctl(
+            s.fileno(),
+            0x8912,  # SIOCGIFCONF
+            struct.pack('iL', bytes, names.buffer_info()[0])
+        ))[0]
+        if outbytes == bytes:
+            max_possible *= 2
+        else:
+            break
+    namestr = names.tostring()
+    return [(namestr[i:i + 16].split('\0', 1)[0],
+             socket.inet_ntoa(namestr[i + 20:i + 24]))
+            for i in range(0, outbytes, struct_size)]
+
+
+def get_interfaces(host_info={}):
+    for interface, ipaddress in all_interfaces():
+        if ipaddress == "127.0.0.1":
+            continue
+        host_info["ipaddress_" + interface] = ipaddress
+        host_info["fqdn_" + interface] = socket.gethostbyaddr(ipaddress)[0]
+    return host_info
+
+
+def get_cpu_info(host_info={}):
+    info_raw = popen(["cat", "/proc/cpuinfo"])
+    model = ""
+    nb_cpu = 0
+    nb_cores = 0
+    nb_units = 0
+    for p in string.split(info_raw, "\n"):
+        if ':' in p:
+            x, y = map(lambda x: string.strip(x), string.split(p, ':', 1))
+            if x.startswith("physical id"):
+                if nb_cpu < int(y):
+                    nb_cpu = int(y)
+            if x.startswith("cpu cores"):
+                if nb_cores < int(y):
+                    nb_cores = int(y)
+            if x.startswith("processor"):
+                if nb_units < int(y):
+                    nb_units = int(y)
+            if x.startswith("model name"):
+                model = y
+
+    nb_cpu += 1
+    nb_units += 1
+    host_info["cpu_model"] = model
+    host_info["physical_cpus"] = str(nb_cpu)
+    host_info["cpu_cores"] = str(nb_cores)
+    host_info["logical_cpus"] = str(nb_units)
+    return host_info
+
+
+# platform module throws exceptions while running within the python collectd plugin
+# shell out to uname instead
+def get_kernel_info(host_info={}):
+    host_info["kernel_name"] = popen(["uname", "-s"])
+    host_info["kernel_release"] = popen(["uname", "-r"])
+    host_info["kernel_version"] = popen(["uname", "-v"])
+    host_info["machine"] = popen(["uname", "-m"])
+    host_info["processor"] = popen(["uname", "-p"])
+    host_info["operating_system"] = popen(["uname", "-o"])
+    return host_info
+
+
+def get_aws(thing, host_info={}):
+    try:
+        r = requests.get("http://169.254.169.254/latest/meta-data/%s" % thing, timeout=0.1)
+        if r.ok:
+            host_info["aws_%s" % thing.replace('-', '_')] = r.text
+    except:
+        pass  # no biggie, probably not on aws
+
+
+def get_aws_info(host_info={}):
+    get_aws("instance-id", host_info)
+    get_aws("instance-type", host_info)
+    get_aws("ami-id", host_info)
+    get_aws("mac", host_info)
+    get_aws("reservation-id", host_info)
+    get_aws("profile", host_info)
```

Comment:
Oh, and rather than getting a bunch of different requests. Look into using http://169.254.169.254/latest/dynamic/instance-identity/document. It's a json file that includes a bunch of identifying info.

Also, and the following is opinion, I am not sure how interesting "mac" or "reservation-id" or "profile" is in the general case. Reservation-ids are the least useful in that they have nothing to do with RIs. They can be used to identify machines that were started at the same time, but I have never needed that info. "Mac" seems not worth the extra request. "Profile" feels the same way to me.